envoy/stream_info/stream_info.h

```diff
@@ -584,16 +584,6 @@ class StreamInfo {
    */
   virtual Tracing::Reason traceReason() const PURE;
 
-  /**
-   * @return Connection ID of the downstream connection, or unset if not available.
-   **/
-  virtual absl::optional<uint64_t> connectionID() const PURE;
```

Comment:
Why not keep `connectionID()` around, but make it a facade for `downstreamAddressProvider().connectionID()` call? This way StreamInfo gets uncoupled from an AddressProvider (see here for an example of such a coupling: https://github.com/envoyproxy/envoy/pull/17231/files#diff-888dc1d9b82ce89abbe565c8c52a11dd16120d2d23b794e1a8c46e5ac0bb039aR817)

Reply:
emm...I'm hesitant when writing the PR. I don't a strong reason to keep the StreamInfo::connectionID or not. In the end, I follow the localAddress()/remoteAddress(), they don't have a facade in StreamInfo. https://github.com/envoyproxy/envoy/blob/a0ca08b69974cc0296e242bfb1c2d4e252d832b5/source/common/formatter/substitution_formatter.cc#L782

If we keep connectionID, I thought it should be able uncouple the calling code and AddressProvider. But AddressProvider interface is quite single, so I think it should be ok. But let me know how you see the benefit of uncouple streaminfo and addressProvider, I will update the PR.

Reply:
I agree with @soulxu. I'm not sure there is any reason to have a facade here?

Reply:
In this PR alone the chain of calls `stream_info.downstreamAddressProvider()...` is used in five places, not counting tests. There are probably going to be more with time? It's not huge for sure, but makes changing and testing the code harder.

Reply:
I'm ok with both ways, if we have a facade for connection id, what about those localaddress/remoteaddress? @mattklein123 what do you think?

Reply:
I agree it doesn't really matter, but all things being equal I don't think it's better to have more methods vs. some slightly longer call chains. I think it's fine as is.